### ISSUE TITLE
Display 'if else' structure better in structure guidelines

### ISSUE THREAD

**Issue by vsfeedback** [dotnet/roslyn#50503]
_This issue has been moved from [a ticket on Developer Community](https://developercommunity2.visualstudio.com/t/Display-if-else-structure-better-in-st/1298987)._

---
[severity:It's more difficult to complete my work]
VS, Code Editor: The window popping up when the mouse rests on a structure guideline should show the previous statements causing the indentation. For the `else` clause, not enough information is shown.

Sample code:

```
using System;

namespace IfLineMissingProblem {
  public class Class1 {
    public Class1() {
      var i = new Random(). Next();
      if (i>0) {
        if (i>1000) {
          if (i>1000000) {

} else {

}
        } else {
          if (i>10) {

} else {

}
        }
      } else {
        if (i<-1000) {
          if (i<-1000000) {

          } else {

          }
        } else {
          if (i>10) {

} else {

}
        }
      }
    }
  }
}
```

When pointing with the mouse on the guideline of the last else clause, the following gets shown:

![image](https://user-images.githubusercontent.com/1408396/104671286-5965f480-5692-11eb-8866-db0f67e5b1df.png)

This is really not helpful. It doesn't help me to understand quickly where in the code I am. It would be much better, if the window would show:

>namespace IfLineMissingProblem
>  public class Class1
>    public Class1()
>      if (i>0)
>      else 
>        if (i<-1000)
>        else 
>          if (i<-10)
>          else

---
#### Feedback Bot on 12/31/2020, 00:43 AM: 

<p>We have directed your feedback to the appropriate engineering team for further evaluation. The team will review the feedback and notify you about the next steps.</p>

#### Feedback Bot on 1/5/2021, 09:41 PM: 

<p>Thank you for sharing your feedback! Our teams prioritize action on product issues with broad customer impact. See details at: <a target='_blank' href="https://docs.microsoft.com/en-us/visualstudio/ide/report-a-problem?view=vs-2019#faq">https://docs.microsoft.com/en-us/visualstudio/ide/report-a-problem?view=vs-2019#faq</a>. In case you need answers to common questions or need assisted support, be sure to use <a target='_blank' href="https://visualstudio.microsoft.com/vs/support/">https://visualstudio.microsoft.com/vs/support/</a>. We’ll keep you posted on any updates to this feedback.</p>

---
### Original Solutions
(no solutions)

**Comment by sharwell** [comment]
The proposed design is to show this:

```patch
 namespace IfLineMissingProblem
   public class Class1
     public Class1()
+      if (i>0)
       else
+        if (...)
         else
+          if (...)
           else
```

**Comment by genlu** [comment]
Design meeting note:
This needs further investigation. @jasonmalinowski

**Comment by jasonmalinowski** [comment]
@gundermanc: so I was looking at this a little bit and I see now that we're on structure tags, there's [IStructureContextSource](https://docs.microsoft.com/en-us/dotnet/api/microsoft.visualstudio.text.adornments.istructurecontextsource.getstructurecontextasync?view=visualstudiosdk-2019) which lets us entirely customize the display here, but we're already handed the structure elements as computed from the internal tree. Is there any API we can implement that lets us specify the elements but keep the standard editor logic for displaying it?

**Comment by gundermanc** [comment]
>but keep the standard editor logic for displaying it?

IBlockTag is a shim around IStructureTag, so the behavior of IStructureTag is a superset. If you start omitting parameters, we revert to the heuristics.

e.g.: if you omit the GuidelineHorizontalAnchorPoint, we'll guess it from the HeaderSpan and the OutliningSpan.

Here's the relevant internal editor sources: https://devdiv.visualstudio.com/DevDiv/_git/VS-Platform?path=%2Fsrc%2FEditor%2FText%2FImpl%2FStructure%2FStructureSpanningTree%2FStructureMappingUtil.cs&version=GBv16.9&line=222&lineEnd=236&lineStartColumn=1&lineEndColumn=70&lineStyle=plain&_a=contents.

```csharp
            // There are three cases to consider for the starting x-point of the block:
            // 1)       if (foo) {
            //          |                           // Line goes here
            //
            // 2a)      if (foo)
            //          {
            //          |                           // Line goes here
            //
            // 2b)      if (foo)
            //              {
            //              |                       // Line goes here
            //
            // 3)       if (bar &&
            //              foo) {
            //          |                           // Line goes here
            // In all cases, find the position of the first non-whitespace character on the line.
```

**Comment by gundermanc** [comment]
The problem in the 'else if' case is that we're being told that the 'header span' starts at the 'i'. Per case 3 above, the guide line gets misplaced, and the tooltip shows 'if' instead of 'else if'.

**Comment by jasonmalinowski** [comment]
@gundermanc Oh don't confuse this with that other conversation. This is for what we show if you mouse over the guideline. Unless I misunderstanding your comment?

**Comment by gundermanc** [comment]
Ohhhh, my mistake :)

>Is there any API we can implement that lets us specify the elements but keep the standard editor logic for displaying it?

Yes, if you don't implement IStructureContextSource, you get our platformized representation of it.

**Comment by jasonmalinowski** [comment]
But there's no way to say "please show this hierarchy, and you do all the projection buffer magic?"

**Comment by gundermanc** [comment]
>But there's no way to say "please show this hierarchy, and you do all the projection buffer magic?"

Are you looking for something other than the below? This is free, simply not implementing IStructureContextSource.

![image](https://user-images.githubusercontent.com/5387680/108579554-491de680-72dc-11eb-8ce9-431db327442a.png)

If you want to customize what your one layer in the hierarchy displays (circled an example) you can do that by changing what you return in IStructureTag.HeaderSpan

![image](https://user-images.githubusercontent.com/5387680/108579585-84201a00-72dc-11eb-9ee5-e288ea1ccba0.png)

If you want to change the entire tooltip contents, you can do that by exporting IStructureContextSource and using the IStructureElements we give you to render the content however you want. Note that this indirection is intentional because mixed language scenarios, like C# in Razor, need tooltips with rendered content from all enclosing languages. Editor does the ~~hard~~ weird work of aggregating structure tags across languages and finding which ones enclose the cursor position.

**Comment by gundermanc** [comment]
TL;DR: adjusting what you give us in 'HeaderSpan' for 'if' and/or 'else' will probably fix the issue.

**Comment by jasonmalinowski** [comment]
@gundermanc: the desire in this case is to show another line entirely. If you are contained within an else block, it'd be nice if the if block line was also shown. I guess what'd we want is some way to say "if when you walked the tags to find the parenting tags, you found this tag, also include another tag the line before."

**Comment by gundermanc** [comment]
>I guess what'd we want is some way to say "if when you walked the tags to find the parenting tags, you found this tag, also include another tag the line before."

If you simply want the tip to show 'else if (CONDITION)' you can do so with the following header span (RED) and outlining span (BLUE):

![image](https://user-images.githubusercontent.com/5387680/108767395-3e956400-750b-11eb-89c7-01c9ca038442.png)

There's unfortunately not currently a way to show non-contiguous spans of text in the tooltip from a single tag, or spans from another document (projection/elision). If there's a strong desire to change this, I could probably tweak the API to let the HeaderSpan reference a span from another document or introduce a version of the structure tag that supports a collection of header spans.

FWIW, you can display any span from the current document in the tooltip, even if it doesn't overlap/intersect with the structure guides or the outlining. If you go this route though, you'll need to populate _all members_ of structure tag.

Each span can be provided separately, so you could simply provide:

- _HeaderSpan_: the span you want in the tooltip. Note that when you partially specify your structure tag, HeaderSpan.Start is what drives the positioning heuristics, so you'll have to fully-specify the structure tag if you move the start position of your header.
- _OutliningSpan_: the span you want to collapse when the user clicks the folding/outlining markers.
- _GuidelineSpan_: a span covering the _vertical_ range of lines that you want the structure guide (gray dotted line) to cover. This determines the Y coordinate and the length of the line.
- _GuidelineHorizontalAnchor_: a point that determines the horizontal offset of the structure guide (the X coordinate). This should probably be on the 'e' in 'else if' OR on the close curly brace, depending on how you want the lines drawn. The idea of this member is that you can easily tie the rendering of your structure guide to a particular code element's position and we'll move it for you.

**Comment by CyrusNajmabadi** [comment]
>GuidelineHorizontalAnchor: a point that determines the horizontal offset of the structure guide (the X coordinate). This should probably be on the 'e' in 'else if' OR on the close curly brace, depending on how you want the lines drawn. The idea of this member is that you can easily tie the rendering of your structure guide to a particular code element's position and we'll move it for you.

Can a helper be provided which answers: which is the leftmost point of these positions?  We want the leftmost part of the region *always*.  This was the heuristic we discussed a long time back that covers all use cases.  If we don't take the leftmost point, we always risk text intersection and things looking 'weird'.

**Comment by jasonmalinowski** [comment]
@gundermanc: but if I had a guideline span going from the start of an `if` to the end of the `else` _and_ a guideline span for just the `else`, what happens? Just assume simple code like:

```
if (blah)
{
}
else
{
}
```

and not any of the fancy indenting cases above.

**Comment by gundermanc** [comment]
>Can a helper be provided which answers: which is the leftmost point of these positions?

Yeah sure. Does this need to be an API or can it be copy-pasted? The code we use to determine this is in editor is pretty trivial:

```csharp
private static int IndentOfLine(ITextSnapshotLine line, int tabsize)
        {
            int indent = 0;

            for (int i = line.Start; (i < line.End); ++i)
            {
                char c = line.Snapshot[i];

                if (c == TabChar)
                {
                    indent = (((indent / tabsize) + 1) * tabsize);
                }
                else if (char.IsWhiteSpace(c))
                {
                    ++indent;
                }
                else
                {
                    return indent;
                }
            }

            return -1;  // Signal nothing but whitespace on the line.
        }
```

>if I had a guideline span going from the start of an if to the end of the else and a guideline span for just the else, what happens?

I think you'll end up with two adornments that partially overlap. This _might_ look ok but it might also mess up the line pattern since the dashes aren't guaranteed to align. Might also be worth checking what the behavior is when doing outlining. Though we would have added tests for this case, Roslyn would be the first to tweak this particular knob...

**Comment by CyrusNajmabadi** [comment]
Closing as external.  This should be handled at the editor layer.
